tardis/visualization/widgets/custom_abundance.py

```diff
@@ -0,0 +1,1700 @@
+                box_output,
+                self.error_view,
+            ]
+        )
+
+    @error_view.capture(clear_output=True)
+    def to_csvy(self, path, overwrite):
+        """Output CSVY file on the specified path.
+
+        Parameters
+        ----------
+            path : str
+                Output path.
+            overwrite : bool
+                True if overwriting, False otherwise.
+        """
+        if not path.endswith(".csvy"):
+            path += ".csvy"
+
+        if os.path.exists(path) and not overwrite:
+            raise FileExistsError(
+                "The file already exists. Click the 'overwrite' checkbox to overwrite it."
+            )
+        else:
+            self.write_yaml_portion(path)
+            self.write_csv_portion(path)
+
+    @error_view.capture(clear_output=True)
+    def write_yaml_portion(self, path):
+        """Write the YAML portion of the output file.
+
+        Parameters
+        ----------
+            path : str
+        """
+        name = path.split("/")[-1]
+        d_time_0 = self.density_editor.density_t_0 * u.day
+        i_time_0 = self.input_i_time_0.value * u.day
+        custom_yaml = CustomYAML(
+            name, d_time_0, i_time_0, self.velocity[0], self.velocity[-1]
+        )
+        custom_yaml.create_fields_dict(self.elements)
+
+        with open(path, "w") as f:
+            yaml_output = yaml.dump(custom_yaml, sort_keys=False)
+
+            # Add YAML delimiter
+            splits = yaml_output.split("\n")
+            splits[0] = splits[-1] = YAML_DELIMITER
+            yaml_output = "\n".join(splits) + "\n"
+
+            f.write(yaml_output)
+
+        print("Saved Successfully!")
+
+    def write_csv_portion(self, path):
+        """Write the CSV portion of the output file.
+
+        Parameters
+        ----------
+            path : str
+        """
+        try:
+            data = self.abundance.T
+            data.columns = self.elements
+            first_row = pd.DataFrame(
+                [[0] * self.no_of_elements], columns=self.elements
+            )
+            data = pd.concat([first_row, data])
+            formatted_v = pd.Series(self.velocity.value).apply(
+                lambda x: "%.3e" % x
+            )
+            density = self.density_editor.d
+            data.insert(0, "velocity", formatted_v)
+            data.insert(1, "density", density)
+
+            data.to_csv(path, mode="a", index=False)
+        except pd.errors.EmptyDataError:
+            data = None
+
+    @classmethod
+    def from_csvy(cls, fpath):
+        """Create a new CustomAbundanceWidget instance with data from CSVY file.
+
+        Parameters
+        ----------
+            fpath : str
+                the path of CSVY file.
+
+        Returns
+        -------
+            CustomAbundanceWidget
+        """
+        csvy_model_config, csvy_model_data = load_csvy(fpath)
+        base_dir = os.path.abspath(os.path.dirname(__file__))
```

Comment:
lets use `tardis.__path__` here as `BASE_DIR` at the beginning of the file.

Reply:
and use `pathlib`

Reply:
@wkerzendorf  I think I can use `pathlib` when writing CSVY output file, but why should we use `pathlib` in `from_csvy` method?